lib/pbench/agent/modules/metaclass.py

```diff
@@ -0,0 +1,106 @@
+from pathlib import Path
+import json
+import os
+
+
+class BadConstruction(Exception):
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __str__(self):
+        return repr(self.msg)
+
+
+class ToolMetadata:
+    def __init__(self, mode, context, logger):
+        self.logger = logger
+        assert mode in (
+            "redis",
+            "json",
+        ), f"Logic bomb! Unexpected mode, {mode}, encountered constructing tool meta data"
+        if not context:
+            raise (BadConstruction("No proper context given"))
+        self.mode = mode
+        if mode == "redis":
+            self.redis_server = context
+            self.json_file = None
+        else:
+            self.redis_server = None
+            json_path = Path(context, "tool-scripts", "meta.json")
+            try:
+                self.json = json_path.resolve(strict=True)
+            except FileNotFoundError:
+                raise Exception(f"missing {json_path}")
+            except Exception:
+                raise
+        self.data = self.__getInitialData()
+
+    def __getInitialData(self):
+        if self.mode == "json":
+            if not os.path.isfile(self.json):
+                self.logger.error(
+                    "There is no tool-scripts/meta.json in given install dir"
+                )
+                return None
+            with self.json.open("r") as json_file:
+                metadata = json.load(json_file)
+        elif self.mode == "redis":
+            try:
+                meta_raw = self.redis_server.get("tool-metadata")
+                if meta_raw is None:
+                    self.logger.error("Metadata has not been loaded into redis yet")
+                    return None
+                meta_str = meta_raw.decode("utf-8")
+                metadata = json.loads(meta_str)
+            except Exception:
+                self.logger.error("Failure to reach redis server")
```

Comment:
Seems like we should refine this a bit to tease out different failure cases.  Perhaps something like:

```
try:
    meta_raw = self.redis_server.get("tool-metadata")
except Exception:
    self.logger.exception("Failure to fetch tool metadata from the Redis server")
    raise
else:
    if meta_raw is None:
        self.logger.error("Metadata has not been loaded into Redis server yet")
        return None
try:
    metadata = json.loads(meta_raw.decode("utf-8"))
except Exception as exc:
    self.logger.error("Bad metadata loaded into Redis server, '%s', json=%r", exc, meta_raw)
    return None